include/swift/Syntax/SyntaxArena.h

```diff
@@ -68,9 +69,24 @@ class SyntaxArena : public llvm::ThreadSafeRefCountedBase<SyntaxArena> {
     auto DidInsert = ChildArenas.insert(Arena);
     if (DidInsert.second) {
       Arena->Retain();
+      assert(!containsReferenceCycle());
     }
   }
 
+  bool containsReferenceCycle(std::set<SyntaxArena *> PreviousNodes = {}) {
+    if (PreviousNodes.count(this) > 0) {
+      return true;
+    }
+    for (auto ChildArena : ChildArenas) {
+      std::set<SyntaxArena *> NewPreviousNodes = PreviousNodes;
+      NewPreviousNodes.insert(this);
+      if (ChildArena->containsReferenceCycle(NewPreviousNodes)) {
+        return true;
+      }
+    }
+    return false;
```

Comment:
Since `PreviousNodes` is already copied, we don't need to make a new copy.

```suggestion
    if (!PreviousNodes.insert(this).second)
      return true;
    return llvm::any_of(ChildArenas, [&](SyntaxArena *Child) {
      Child->containsReferenceCycle(PreviousNodes)
    });
```

Reply:
Oh, yeah. For some reason I though I was adding a different node depending on which child node I was visiting. But of course I’m not. Also your implementation looks much cleaner. I changed it. Also renamed `PreviousNodes` to `VisitedArenas`.